syntax = "proto3";

package currency;

// Mensajes para conversión normal
message ConvertRequest {
  string from_currency = 1; 
  string to_currency = 2;   
  double amount = 3;        
}

message ConvertReply {
  double converted_amount = 1;
  double rate = 2; 
  string from_currency = 3;
  string to_currency = 4;
}

message Currency {
  string code = 1; 
  string name = 2; 
}

message Empty {}

// Mensajes para el DESAFÍO (GetRate)
message RateRequest {
  string from_currency = 1;
  string to_currency = 2;
}

message RateReply {
  double rate = 1;
}

// Definición del Servicio
service CurrencyConverter {
  // 1. Conversión simple
  rpc Convert(ConvertRequest) returns (ConvertReply);

  // 2. Lista de monedas
  rpc GetSupportedCurrencies(Empty) returns (stream Currency);

  // 3. Stream de tasas
  rpc StreamRates(Empty) returns (stream ConvertReply);

  // 4. NUEVO: Obtener solo la tasa (El desafío)
  rpc GetRate(RateRequest) returns (RateReply);
}